### PR TITLE
feature(generator): add support to detect out of date generated code and build to generate code

### PR DESCRIPTION
**Pull request by scotthart**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5759)

### CODE REVIEW

**Comment by scotthart** [comment]
At this point in the PR, all bazel based presubmit tests should fail because of a googleapis commit hash conflict between the bazel dependencies and the existing generated code in iam and logging.

File: ci/kokoro/docker/Dockerfile.ubuntu [hunk 1]
```diff
@@ -23,6 +23,7 @@ RUN apt-get update && \
         ccache \
         clang \
         clang-9 \
+        clang-format-9 \
```
Comment (coryan) [inline ci/kokoro/docker/Dockerfile.ubuntu:26]: Ah, I see why you did this.  I still it is better to run this build in the same distro that runs the formatting.  Soon (around 2021-04, maybe 2021-10) we will update the `clang-tidy` build to the next Fedora, get a new `clang-format` with it, and there is no guarantee you will be able to find the same version in Ubuntu:18.04 (or Ubuntu:20.04).

File: ci/kokoro/docker/Dockerfile.ubuntu [hunk 2]
```diff
@@ -49,6 +50,9 @@ RUN apt-get update && \
         ca-certificates \
         apt-transport-https
 
+# Provide access to clang-format in a well known place.
+RUN ln -s /usr/bin/clang-format-9 /usr/local/bin/clang-format
```
Comment (coryan) [inline ci/kokoro/docker/Dockerfile.ubuntu:54]: Normally one would use `update-alternatives` for this, see below.

**Review by coryan — CHANGES_REQUESTED** [review]
Some thoughts below, I may not be the best person to do this.

Separate thought, we have at least 2 of these "update-xxx" builds now.  Seems like we need to document this in the team documentation for releases or something like that.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5759)* status: 0 of 6 files reviewed, 6 unresolved discussions (waiting on @coryan and @scotthart)

___
*[ci/kokoro/docker/build.sh, line 117 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5759#-MS4hlvcL9UMSDvso1Xd-r2-117:-MS4sSq7DbKHTXLjPAf1:b-pasova) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d74fc52700e6fc14d49a3ceadfde3427921c407a/ci/kokoro/docker/build.sh#L117)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

That seems unrelated to what the `generate-libraries` build does?  Why not just remove them?
</blockquote></details>

The only tests that exist to run against a generated library are its integration tests. I felt it would be useful when regenerating the library code from updated proto files to compile and execute the integration tests to immediately determine if there are any API and/or behavioral issues.
___
*[ci/kokoro/docker/Dockerfile.ubuntu, line 26 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5759#-MS4hlvcL9UMSDvso1Xb-r2-26:-MS4tF8_2pa8q6M_kCvm:b-dqgxok) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d74fc52700e6fc14d49a3ceadfde3427921c407a/ci/kokoro/docker/Dockerfile.ubuntu#L26)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Ah, I see why you did this.  I still it is better to run this build in the same distro that runs the formatting.  Soon (around 2021-04, maybe 2021-10) we will update the `clang-tidy` build to the next Fedora, get a new `clang-format` with it, and there is no guarantee you will be able to find the same version in Ubuntu:18.04 (or Ubuntu:20.04).
 
</blockquote></details>

So change the generate-libraries build to use the same fedora version as clang-tidy?

**Review by coryan — CHANGES_REQUESTED** [review]
> Reviewed 5 of 6 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5759)* status: all files reviewed, 6 unresolved discussions (waiting on @scotthart)

___
*[ci/kokoro/docker/build.sh, line 114 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5759#-MS4hlvcL9UMSDvso1Xd-r2-114:-MS4v2KY6yaQngYfdMb1:bnx6uen) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d74fc52700e6fc14d49a3ceadfde3427921c407a/ci/kokoro/docker/build.sh#L114)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

That is a pretty old version of ubuntu by now (2018-04, almost 3 years).  Consider using 20.04.
</blockquote></details>

Maybe not... changing to Fedora sounds better.
___
*[ci/kokoro/docker/build.sh, line 117 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5759#-MS4hlvcL9UMSDvso1Xd-r2-117:-MS4uyRn6AqI1g_UDmqG:b9qpg5b) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d74fc52700e6fc14d49a3ceadfde3427921c407a/ci/kokoro/docker/build.sh#L117)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

The only tests that exist to run against a generated library are its integration tests. I felt it would be useful when regenerating the library code from updated proto files to compile and execute the integration tests to immediately determine if there are any API and/or behavioral issues.
</blockquote></details>

Ah, I see now. Makes sense.
___
*[ci/kokoro/docker/Dockerfile.ubuntu, line 26 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5759#-MS4hlvcL9UMSDvso1Xb-r2-26:-MS4v7Z3030WKArV2lK1:bb7sa0c) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d74fc52700e6fc14d49a3ceadfde3427921c407a/ci/kokoro/docker/Dockerfile.ubuntu#L26)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

So change the generate-libraries build to use the same fedora version as clang-tidy?
</blockquote></details>

yes.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5759)* status: all files reviewed, 6 unresolved discussions (waiting on @coryan)

___
*[ci/kokoro/docker/build.sh, line 114 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5759#-MS4hlvcL9UMSDvso1Xd-r2-114:-MS59SR87e4EXtnHvxI0:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d74fc52700e6fc14d49a3ceadfde3427921c407a/ci/kokoro/docker/build.sh#L114)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Maybe not... changing to Fedora sounds better.
</blockquote></details>

Done.
___
*[ci/kokoro/docker/build.sh, line 117 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5759#-MS4hlvcL9UMSDvso1Xd-r2-117:-MS59TBI84v-HPce3wtW:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d74fc52700e6fc14d49a3ceadfde3427921c407a/ci/kokoro/docker/build.sh#L117)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Ah, I see now. Makes sense.
</blockquote></details>

Done.
___
*[ci/kokoro/docker/Dockerfile.ubuntu, line 26 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5759#-MS4hlvcL9UMSDvso1Xb-r2-26:-MS59Ty9BUroaqbpue_M:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d74fc52700e6fc14d49a3ceadfde3427921c407a/ci/kokoro/docker/Dockerfile.ubuntu#L26)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

yes.
</blockquote></details>

Done.
___
*[ci/kokoro/docker/Dockerfile.ubuntu, line 54 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5759#-MS4hlvcL9UMSDvso1Xb-r2-54:-MS59Uam05vSgh-v-PmS:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d74fc52700e6fc14d49a3ceadfde3427921c407a/ci/kokoro/docker/Dockerfile.ubuntu#L54)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Normally one would use `update-alternatives` for this, see below.
</blockquote></details>

Done.
___
*[generator/generate-libraries.sh, line 48 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5759#-MS4hlvcL9UMSDvso1Xg-r2-48:-MS59VPD0ncBVzCbjy98:bftvy31) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d74fc52700e6fc14d49a3ceadfde3427921c407a/generator/generate-libraries.sh#L48)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

If it makes sense to always log from the generator, maybe set the logging on programmatically in `main()`?
</blockquote></details>

At this point in time, I'm not sure logging always needs to be on, but I will keep it mind.
___
*[generator/generate-libraries.sh, line 60 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5759#-MS4hlvcL9UMSDvso1Xg-r2-60:-MS59cHd9UxnCKh7ApG4:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d74fc52700e6fc14d49a3ceadfde3427921c407a/generator/generate-libraries.sh#L60)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

The filename seems yucky to me.  We hide other similar stuff in `ci/` would you consider doing the same?
</blockquote></details>

Done.

**Review by coryan — APPROVED** [review]
> Reviewed 5 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5759)* status: all files reviewed, 1 unresolved discussion (waiting on @scotthart)

___
*[ci/etc/generator-googleapis-commit-hash.sh, line 1 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5759#-MS5K6ye2j5EDTBb_Wyl:-MS5K6ye2j5EDTBb_Wym:b-ifib5y) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1db94d359bf7bb91d89bb392b81ece4950f96852/ci/etc/generator-googleapis-commit-hash.sh#L1)):*
> ```Shell
> #!/usr/bin/env bash
> ```

nit: you may not need the boilerplate, your call.

**Review by coryan — APPROVED** [review]
Looks good, but you need to get the builds to pass, of course.

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5759#-:-MS5KZCR-Apz3A1ekcFS:bwwpq33)* status: all files reviewed, 1 unresolved discussion (waiting on @scotthart)

**Pull request merged**